rpc/client.go

```diff
@@ -1,33 +1,72 @@
 package rpc
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
 	"runtime"
+	"strconv"
 	"sync"
+	"time"
 
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
 	"github.com/evergreen-ci/juniper/gopb"
 	"github.com/evergreen-ci/poplar"
 	"github.com/evergreen-ci/poplar/rpc/internal"
+	"github.com/google/uuid"
 	"github.com/mongodb/grip"
 	"github.com/mongodb/grip/message"
 	"github.com/mongodb/grip/recovery"
 	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 )
 
+// UploadReportOptions contains all the required options for uploading a report
+// to both Cedar and the S3 data-pipes service.
 type UploadReportOptions struct {
 	Report          *poplar.Report
 	ClientConn      *grpc.ClientConn
 	SerializeUpload bool
+	AWSAccessKey    string
+	AWSSecretKey    string
+	AWSToken        string
+	DataPipesHost   string
+	DataPipesRegion string
 	DryRun          bool
 }
 
+// SignedUrl is a struct representing a signed url returned from the data pipes API.
+type SignedUrl struct {
+	SignedUrl      string `json:"signed_url"`
+	ExpirationSecs int    `json:"expiration_secs"`
+}
+
+// UploadReport does the following:
+// 1. Ingest the report.json.
+// 2. Send all artifact files to the user-specified buckets using user-specified keys.
+// 3. Send the metrics metadata and pre-calculated summaries to Cedar over gRPC using the Cedar creds.
+// 4. Send the metrics metadata and pre-calculated summaries to the data-pipes service using AWS keys.
 func UploadReport(ctx context.Context, opts UploadReportOptions) error {
+	var returnError error
```

Comment:
Instead of doing this if/else logic you can just keep it as it was and put the call to `uploadResultsToDataPipes` in a defer statement so it is always triggered when the function returns.

```
defer func() {
        if err := uploadResultsToDataPipes(...), err != nil {
                grip.Warning(message.WrapError(err, message.Fields{
                        "op": "uploadResultsToDataPipes"
                })
        }
}()
```

Reply:
ooh nice. Defer is one of the coolest parts of golang imho